test/Autofac.Test/ModuleTests.cs

```diff
@@ -59,6 +59,42 @@ public void AttachesToRegistrations()
             Assert.Equal(container.ComponentRegistry.Registrations.Count(), attachingModule.Registrations.Count);
         }
 
+        [Fact]
+        public void AttachesToRegistrationsInScope()
+        {
+            var attachingModule = new AttachingModule();
+            Assert.Equal(0, attachingModule.Registrations.Count);
+
+            var builder = new ContainerBuilder();
+            builder.RegisterModule(attachingModule);
+
+            using (var container = builder.Build())
+            using (var scope = container.BeginLifetimeScope(c => c.RegisterType(typeof(int))))
+            {
+                var expected = container.ComponentRegistry.Registrations.Count() + scope.ComponentRegistry.Registrations.Count();
+                Assert.Equal(expected, attachingModule.Registrations.Count);
+            }
+        }
+
+        [Fact]
+        public void AttachesToRegistrationsInNestedScope()
+        {
+            var attachingModule = new AttachingModule();
+            Assert.Equal(0, attachingModule.Registrations.Count);
+
+            var builder = new ContainerBuilder();
+            builder.RegisterModule(attachingModule);
+
+            using (var container = builder.Build())
+            using (var outerScope = container.BeginLifetimeScope(c => c.RegisterType(typeof(int))))
+            using (var innerScope = outerScope.BeginLifetimeScope(c => c.RegisterType(typeof(double))))
+            {
+                var expected = container.ComponentRegistry.Registrations.Count()
+                    + outerScope.ComponentRegistry.Registrations.Count() + innerScope.ComponentRegistry.Registrations.Count();
```

Comment:
Should we test that the module _removes_ its subscriptions as scopes are disposed?

Reply:
But why would it need to? Once the scope gets out of scope (pun intended) and becomes unreachable module's registration will became irrelevant and waiting GC together will scope's component registry. Or am I missing something?

Reply:
I just want to avoid multigenerational garbage collection problems (ie some event handler inadvertently keeping references alive) and/or the `Properties[MetadataKeys.RegisteredPropertyKey]` set of registrations becoming so large it becomes problematic. The whole way this works is because the properties dictionary is inherited from parent to child, right? And as it reads-through to the parent scope, it's actually attaching the new set of event handlers to the object reference in the parent, not something scoped to only the current scope. Wouldn't that mean if I fire up 100 child lifetime scopes I'd be adding to the subscription list and never removing them? Would that cause problems in that the delegate chain in the event handler would become invalid?

I'm not entirely sure, it just occurred to me that we might have a leak of some nature here given there's no cleanup. I've had some experience with that - I tried to implement a thing that would track child lifetime scopes in an effort to dispose of them when a parent was disposed, did something sort of similar to this... and found that I accidentally ended up keeping references to every child scope so nothing got disposed. It was... not good. ;)

Reply:
Yes, but in this instance module is not attaching to component registry - only to possibly _new registrations_ and those will be brand new and nothing to do with reference in the parent.

Only if you try adding new handler to `Registered` inside new scope you'd be looking at the original delegate from the parent registry - but the moment you attempt to modify it new properties dictionary will be created with original values copied (well, sort of) and since delegates are immutable the parent and child will end up with their distinct versions of it. See the unit test I just added.